### ISSUE TITLE
Unify class agnostic and class specific region classification and regression in multi-task learning

### ISSUE THREAD

**Issue by futurely** [rbgirshick/py-faster-rcnn#16]
Fast R-CNN trains the final sibling fully connected layers for object class aware regions classification and regression using mini-batches of region proposals. RPN trains similar sibling fc layers for object vs. non-object exhaustively on multiple anchor boxes at each position of the final conv layer feature maps. One of the aims of Fast R-CNN is to unify the multi-stage training of previous methods. But RPN uses a 4-step alternating training scheme to share conv layers with the Fast R-CNN. Isn't that a step going backwards? The situation is even more evident in the special case of 1-class object detection such as https://github.com/rbgirshick/py-faster-rcnn/issues/10.

If the classification and regression tasks in both Fast R-CNN and RPN can share the conv layers in multi-task learning settings, why can't the four tasks be trained together? The final four sibling fc layers can all use the dense anchor boxes as input collapsing the two-step cascade into a single step and eliminating the separate RPN training and testing completely which is exactly the prediction in the conclusion of the Fast R-CNN paper.

@rbgirshick, @ShaoqingRen what're you opinions on the plausibility?

**Comment by futurely** [comment]
RPN questioned the possibility of joint training. "Both RPN and Fast R-CNN, trained independently, will modify their conv layers in different ways. We therefore need to develop a technique that allows for sharing conv layers between the two networks, rather than learning two separate networks. Note that this is not as easy as simply defining a single network that includes both RPN and Fast R-CNN, and then optimizing it jointly with backpropagation. The reason is that Fast R-CNN training depends on fixed object proposals and it is not clear a priori if learning Fast R-CNN while simultaneously changing the proposal mechanism will converge. While this joint optimizing is an interesting question for future work, we develop a pragmatic 4-step training algorithm to learn shared features via alternating optimization."

BoxSup [1] describes how to change the labeling of candidate segments and update the network parameters simultaneously in an end-to-end iterative training algorithm showing that RPN's doubt about the convergence can be resolved. The candidate segments in BoxSup is the counterpart of the anchor boxes in RPN. The only difference is that the segments are initialized with selective search while the anchor boxes are initialized with the ground truth bounding boxes. Therefore, RPN, Fast R-CNN and BoxSup have the potential to be unified in a single network and achieve state-of-the-art object detection and semantic segmentation at the same time [2].

[1] Jifeng Dai, Kaiming He, and Jian Sun. BoxSup: Exploiting Bounding Boxes to Supervise Convolutional Networks for Semantic Segmentation. ICCV 2015.
[2] Ziwei Liu, Xiaoxiao Li, Ping Luo, Chen Change Loy, Xiaoou Tang. Semantic Image Segmentation via Deep Parsing Network. ICCV 2015.

**Comment by rbgirshick** [comment]
@futurely it is actually very easy to train RPN + Fast R-CNN approximately* jointly. I'm planning on adding prototxt files and training scripts for how to do this soon. The mAP ends up about the same, but the training time is reduced by about 50%.

* "approximately" because the RoI pooling layer is not differentiable w.r.t. the RoI coordinates and therefore gradients of the objective w.r.t. the RoI coordinates cannot be passed back through the network. This is likely a minor issue.

**Comment by futurely** [comment]
That will further accelerate the experiments iterative cycles and enable many more insightful discoveries. Thanks a lot for sharing!

**Comment by ShaoqingRen** [comment]
@rbgirshick Thanks for the detail explanation.
@futurely Joint training is resonable for RPN and Fast R-CNN.
RPN, Fast R-CNN and segmentation can also be joint trained considering the backpropagate of RoI coordinates. Jifeng Dai and Kaiming He have submitted a paper about this.

**Comment by futurely** [comment]
It's great to hear about another break through in this area. Practical applications are benefiting a lot from the latest algorithms.

**Comment by marphil** [comment]
@rbgirshick I think the whole ecosystem would benefit from the prototxt files and training scripts! Can't wait to get my hands on them!

**Comment by futurely** [comment]
Fast R-CNN and RPN mainly differ in the sources of the input rois. The prototxt files and the source code reflect that the ROIDataLayer shared by both networks and the AnchorTargetLayer specific to RPN basically do the same things. They both output labels used by the SoftmaxWithLoss layer, bbox_targets, bbox_inside_weights and bbox_outside_weights used by the SmoothL1Loss layer. Their internal code path are highly similar except that the calling trace of the ROIDataLayer is a bit long. It would be much easier to unify the implementations if the rois logics are separated into classes inheriting from the same base class.

|data | ROIDataLayer | AnchorTargetLayer |
| ------------- |:-------------:| -----:|
| roidb | RoIDataLayer.set_roidb fast_rcnn.train.SolverWrapper.\__init__  tools.train_net.train_net tools.train_net.combined_roidb fast_rcnn.train.get_training_roidb(imdb) roi_data_layer.roidb.prepare_roidb(imdb) tools.train_net.combined_roidb.get_imdb datasets.factory.get_imdb datasets.pascal_voc | bottom[*].data |
| labels | RoIDataLayer.forward ROIDataLayer._get_next_minibatch roi_data_layer.minibatch.get_minibatch | generate_anchors.generate_anchors |
| bbox_targets | roi_data_layer.roidb._compute_targets | rpn.anchor_target_layer._compute_targets |
| bbox_inside_weights | RoIDataLayer.forward RoIDataLayer._get_next_minibatch roi_data_layer.minibatch.get_minibatch roi_data_layer.minibatch._sample_rois roi_data_layer.minibatch._get_bbox_regression_labels cfg.TRAIN.BBOX_INSIDE_WEIGHTS | rpn.anchor_target_layer._unmap |
| bbox_outside_weights | np.array(bbox_inside_blob > 0).astype(np.float32) | rpn.anchor_target_layer._unmap |

fast_rcnn_train.pt
```
layer {
  name: 'data'
  type: 'Python'
  top: 'data'
  top: 'rois'
  top: 'labels'
  top: 'bbox_targets'
  top: 'bbox_inside_weights'
  top: 'bbox_outside_weights'
  python_param {
    module: 'roi_data_layer.layer'
    layer: 'RoIDataLayer'
  }
}
...
layer {
  name: "loss_cls"
  type: "SoftmaxWithLoss"
  bottom: "cls_score"
  bottom: "labels"
}
layer {
  name: "loss_bbox"
  type: "SmoothL1Loss"
  bottom: "bbox_pred"
  bottom: "bbox_targets"
  bottom: "bbox_inside_weights"
  bottom: "bbox_outside_weights"
}
```
rpn_train.pt
```
layer {
  name: 'input-data'
  type: 'Python'
  top: 'data'
  top: 'im_info'
  top: 'gt_boxes'
  python_param {
    module: 'roi_data_layer.layer'
    layer: 'RoIDataLayer'
  }
}
...
layer {
  name: 'rpn-data'
  type: 'Python'
  bottom: 'rpn_cls_score'
  bottom: 'gt_boxes'
  bottom: 'im_info'
  bottom: 'data'
  top: 'rpn_labels'
  top: 'rpn_bbox_targets'
  top: 'rpn_bbox_inside_weights'
  top: 'rpn_bbox_outside_weights'
  python_param {
    module: 'rpn.anchor_target_layer'
    layer: 'AnchorTargetLayer'
  }
}
...
layer {
  name: "rpn_loss_cls"
  type: "SoftmaxWithLoss"
  bottom: "rpn_cls_score_reshape"
  bottom: "rpn_labels"
}
layer {
  name: "rpn_loss_cls"
  type: "SoftmaxWithLoss"
  bottom: "rpn_cls_score_reshape"
  bottom: "rpn_labels"
}
layer {
  name: "rpn_loss_bbox"
  type: "SmoothL1Loss"
  bottom: "rpn_bbox_pred"
  bottom: "rpn_bbox_targets"
  bottom: "rpn_bbox_inside_weights"
  bottom: "rpn_bbox_outside_weights"
}
```

**Comment by futurely** [comment]
The end-to-end approximate joint training has been released. https://github.com/rbgirshick/py-faster-rcnn/compare/81ee227...28a7c0b
